ycsb: Fix rate limiting and resetting of stats after loading phase

a-robinson opened the pull request.

a-robinson committed:
ycsb: Fix rate limiting mechanism to avoid overly limiting

rate.Limiter does the right thing for us. This is the same rate limiting
mechanism that kv uses. It takes us from output like:

Loading complete: 7.0s
     0s       1305.9           6 / 0 / 0        1301 / 0           0 / 0
     1s        128.7         124 / 0 / 0           5 / 0           0 / 0
     2s        128.8         125 / 0 / 0           4 / 0           0 / 0
     3s        131.9         124 / 0 / 0           8 / 0           0 / 0
     4s        132.3         124 / 0 / 0           8 / 0           0 / 0
     5s        132.4         126 / 0 / 0           6 / 0           0 / 0

To output like:

Loading complete: 6.5s
     0s 18470906513000927232.0         136 / 0 / 0 18446744073709542423 / 0           0 / 0
     1s        299.4         283 / 0 / 0          17 / 0           0 / 0
     2s        300.6         285 / 0 / 0          15 / 0           0 / 0
     3s        299.4         286 / 0 / 0          14 / 0           0 / 0
     4s        300.0         288 / 0 / 0          12 / 0           0 / 0
     5s        300.1         283 / 0 / 0          17 / 0           0 / 0

When running `./ycsb --concurrency=3 --rate-limit=100`

Fixes #119

a-robinson committed:
ycsb: Fix resetting of stats after data loading is done

Takes us from output like:

Loading complete: 6.5s
     0s 18470906513000927232.0         136 / 0 / 0 18446744073709542423 / 0           0 / 0
     1s        299.4         283 / 0 / 0          17 / 0           0 / 0
     2s        300.6         285 / 0 / 0          15 / 0           0 / 0
     3s        299.4         286 / 0 / 0          14 / 0           0 / 0
     4s        300.0         288 / 0 / 0          12 / 0           0 / 0
     5s        300.1         283 / 0 / 0          17 / 0           0 / 0

To output like:

Loading complete: 6.8s
     0s        297.6          48 / 0 / 0           6 / 0           0 / 0
     1s        300.0         282 / 0 / 0          18 / 0           0 / 0
     2s        300.0         283 / 0 / 0          17 / 0           0 / 0
     3s        299.9         281 / 0 / 0          19 / 0           0 / 0
     4s        300.1         285 / 0 / 0          15 / 0           0 / 0
     5s        299.5         283 / 0 / 0          17 / 0           0 / 0

When running `./ycsb --concurrency=3 --rate-limit=100`

a-robinson commented:
Combined, these two changes take us from post-data-loading output like this:

```
Loading complete: 7.0s
     0s       1305.9           6 / 0 / 0        1301 / 0           0 / 0
     1s        128.7         124 / 0 / 0           5 / 0           0 / 0
     2s        128.8         125 / 0 / 0           4 / 0           0 / 0
     3s        131.9         124 / 0 / 0           8 / 0           0 / 0
     4s        132.3         124 / 0 / 0           8 / 0           0 / 0
     5s        132.4         126 / 0 / 0           6 / 0           0 / 0
```

To output like this:

```
Loading complete: 6.8s
     0s        297.6          48 / 0 / 0           6 / 0           0 / 0
     1s        300.0         282 / 0 / 0          18 / 0           0 / 0
     2s        300.0         283 / 0 / 0          17 / 0           0 / 0
     3s        299.9         281 / 0 / 0          19 / 0           0 / 0
     4s        300.1         285 / 0 / 0          15 / 0           0 / 0
     5s        299.5         283 / 0 / 0          17 / 0           0 / 0
```

When running `./ycsb --concurrency=3 --rate-limit=100`

petermattis commented on ycsb/main.go:
```diff
@@ -74,7 +76,7 @@ var writeDuration = flag.Duration("write-duration", 0,
 	"Drop the existing table and recreate it to start from scratch")
-var rateLimit = flag.Uint64("rate-limit", 0,
+var maxRate = flag.Uint64("rate-limit", 0,
 	"Maximum number of operations per second per worker. Set to zero for no rate limit")
```
In `kv`, the rate limit is global, not per worker. We should probably do the same here for consistency. I think you just have to pass in a shared `*rate.Limiter` to `newYcsbWorker`.

rjnn replied:
+1

a-robinson replied:
Done.

petermattis reviewed (approved):
LGTM

rjnn reviewed (approved):
LGTM

a-robinson committed:
ycsb: Switch from per-client rate limiting to per-process limiting

Also changed the name of the flag to match the identical flag in kv.
Normally I wouldn't change a flag name, but given that this is also a
big change in behavior it seems like a good time to move it.

a-robinson commented:
Comment addressed. I also changed the name of the flag to match the identical flag in kv. Normally I wouldn't change a flag name, but given that this is also a big change in behavior it seems like a good time to move it.
